File: Zeze/Transaction/Collections/PSet1.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;

namespace Zeze.Transaction.Collections
{
    public sealed class PSet1<E> : PSet<E>
    {
        public PSet1(long logKey, Func<ImmutableHashSet<E>, Log> logFactory) : base(logKey, logFactory)
        {
        }

        public override bool Add(E item)
        {
            if (item == null)
                throw new ArgumentNullException();

            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : set;
                var newv = oldv.Add(item);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ((ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteSet<E>(this))).LogAdd(item);
                    return true;
                }
                else
                {
                    return false;
                }
            }
            else
            {
                set = set.Add(item);
                return true;
            }
        }

        public override void Clear()
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : set;
                if (!oldv.IsEmpty)
                {
                    txn.PutLog(NewLog(ImmutableHashSet<E>.Empty));
                    ChangeNoteSet<E> note = (ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteSet<E>(this));
                    foreach (var item in oldv)
                        note.LogRemove(item);
                }
            }
            else
            {
                set = ImmutableHashSet<E>.Empty;
            }
        }



        public override void ExceptWith(IEnumerable<E> other)
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : set;
                var newv = oldv.Except(other);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ChangeNoteSet<E> note = (ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteSet<E>(this));
                    foreach (var item in other)
                        note.LogRemove(item);
                }
            }
            else
            {
                set = set.Except(other);
            }
        }


        public override void IntersectWith(IEnumerable<E> other)
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : set;
                var newv = oldv.Intersect(other);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ChangeNoteSet<E> note = (ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteSet<E>(this));
                    foreach (var old in oldv)
                    {
                        if (false == other.Contains(old))
                            note.LogRemove(old);
                    }
                }
            }
            else
            {
                set = set.Intersect(other);
            }
        }

        public override bool Remove(E item)
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : set;
                var newv = oldv.Remove(item);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ((ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteSet<E>(this))).LogRemove(item);
                    return true;
                }
                else
                {
                    return false;
                }
            }
            else
            {
                var old = set;
                set = set.Remove(item);
                return old != set;
            }
        }

        public override void SymmetricExceptWith(IEnumerable<E> other)
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : set;
                var newv = oldv.SymmetricExcept(other);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ChangeNoteSet<E> note = (ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteSet<E>(this));
                    // this: 1,2 other: 2,3 result: 1,3
                    foreach (var item in other)
                    {
                        if (oldv.Contains(item))
                        {
                            note.LogRemove(item);
                        }
                        else
                        {
                            note.LogAdd(item);
                        }
                    }
                }
            }
            else
            {
                set = set.SymmetricExcept(other);
            }
        }

        public override void UnionWith(IEnumerable<E> other)
        {
            foreach (E v in other)
            {
                if (null == v)
                    throw new ArgumentNullException();
            }

            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(LogKey) is LogV log ? log.Value : set;
                var newv = oldv.Union(other);
                if (newv != oldv)
                {
                    txn.PutLog(NewLog(newv));
                    ChangeNoteSet<E> note = (ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.ObjectId, () => new ChangeNoteSet<E>(this));
                    foreach (var item in other)
                    {
                        if (false == oldv.Contains(item))
                            note.LogAdd(item);
                    }
                }
            }
            else
            {
                set = set.Union(other);
            }
        }

        protected override void InitChildrenRootInfo(Record.RootInfo tableKey)
        {
        }
    }
}

File: ZezeJava/src/Zeze/Transaction/Collections/PSet1.java
package Zeze.Transaction.Collections;

import java.util.Collection;
import org.pcollections.Empty;
import Zeze.Transaction.*;

public final class PSet1<E> extends PSet<E> {
	public PSet1(long logKey, LogFactory<org.pcollections.PSet<E>> logFactory) {
		super(logKey, logFactory);
	}

	@SuppressWarnings("unchecked")
	@Override
	public boolean add(E item) {
		if (item == null) {
			throw new NullPointerException();
		}

		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			var olds = null != log ? ((LogV<E>)log).Value : set;
			var news = olds.plus(item);
			if (news != olds) {
				txn.PutLog(NewLog(news));
				((ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.getObjectId(), () -> new ChangeNoteSet<E>(this))).LogAdd(item);
				return true;
			}
			else {
				return false;
			}
		}
		else {
			var olds = set;
			set = set.plus(item);
			return olds != set;
		}
	}

	@Override
	public void clear() {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var olds = null != log ? ((LogV<E>)log).Value : set;
			if (!olds.isEmpty()) {
				txn.PutLog(NewLog(Empty.set()));
				@SuppressWarnings("unchecked")
				ChangeNoteSet<E> note = (ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.getObjectId(), () -> new ChangeNoteSet<E>(this));
				for (var item : olds) {
					note.LogRemove(item);
				}
			}
		}
		else {
			set = Empty.set();
		}
	}


	@SuppressWarnings("unchecked")
	@Override
	public boolean remove(Object item) {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			var olds = null != log ? ((LogV<E>)log).Value : set;
			var news = olds.minus(item);
			if (news != olds) {
				txn.PutLog(NewLog(news));
				((ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.getObjectId(), () -> new ChangeNoteSet<E>(this))).LogRemove((E)item);
				return true;
			}

			return false;
		}
		else {
			var old = set;
			set = set.minus(item);
			return old != set;
		}
	}

	@Override
	protected void InitChildrenRootInfo(Zeze.Transaction.Record.RootInfo tableKey) {
	}

	@Override
	public boolean addAll(Collection<? extends E> c) {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var olds = null != log ? ((LogV<E>)log).Value : set;
			var news = olds.plusAll(c);
			if (news != olds) {
				txn.PutLog(NewLog(news));
				@SuppressWarnings("unchecked")
				var note = ((ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.getObjectId(), () -> new ChangeNoteSet<E>(this)));
                for (var item : c) {
                    if (false == olds.contains(item))
                        note.LogAdd(item);
                }
			}
		}
		else {
			set = set.plusAll(c);
		}
		return true;
	}

	@Override
	public boolean removeAll(Collection<?> c) {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var olds = null != log ? ((LogV<E>)log).Value : set;
			var news = olds.minusAll(c);
			if (news != olds) {
				txn.PutLog(NewLog(news));
				@SuppressWarnings("unchecked")
				var note = ((ChangeNoteSet<E>)txn.GetOrAddChangeNote(this.getObjectId(), () -> new ChangeNoteSet<E>(this)));
                for (var item : c) {
                    note.LogRemove((E)item);
                }
                return true;
			}
			return false;
		}
		else {
			var olds = set;
			set = set.minusAll(c);
			return olds != set;
		}
	}

}